kvcoord: Eliminate 1 Go routine from MuxRangeFeed

miretskiy opened the pull request:
Prior to this PR, the server side `MuxRangeFeed`
implementation spawned a separate Go routine executing
single RangeFeed for each incoming request.

This is wasteful and unnecessary.
Instead of blocking, and waiting for a single RangeFeed to complete,
have rangefeed related functions return a promise to return
a `*roachpb.Error` once rangefeed completes (`future.Future[*roachpb.Error]`).

Prior to this change MuxRangeFeed would spin up 4 Go routines
per range.  With this PR, the number is down to 3.
This improvement is particularly important when executing
rangefeed against large tables (10s-100s of thousands of ranges).

Informs #96395
Epic: None

Release note (enterprise change): Changefeeds running with
`changefeed.mux_rangefeed.enabled` setting set to true are
more efficient, particularly when executing against large tables.

Commit:
util: Add a `Future[T]` library

Add a library implementing promise/future abstraction.
`Future[T]` describes a value of type T, which will become
available in the future.  The caller then may wait until
future becomes available.

Release note: None

Commit:
kvcoord: Eliminate 1 Go routine from MuxRangeFeed

Prior to this PR, the server side `MuxRangeFeed`
implementation spawned a separate Go routine executing
single RangeFeed for each incoming request.

This is wasteful and unnecessary.
Instead of blocking, and waiting for a single RangeFeed to complete,
have rangefeed related functions return a promise to return
a `*kvpb.Error` once rangefeed completes (`future.Future[*kvpb.Error]`).

Prior to this change MuxRangeFeed would spin up 4 Go routines
per range.  With this PR, the number is down to 3.
This improvement is particularly important when executing
rangefeed against large tables (10s-100s of thousands of ranges).

Informs #96395
Epic: None

Release note (enterprise change): Changefeeds running with
`changefeed.mux_rangefeed.enabled` setting set to true are
more efficient, particularly when executing against large tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96756)
<!-- Reviewable:end -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96756)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @erikgrinaker, @irfansharif, @miretskiy, and @tbg)

___
*[`-- commits` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96756#-NNiMhvE43U2kKAKVCBh):*
drive by: I don't think this is a release note that users can understand :P


<!-- Sent from Reviewable.io -->

miretskiy commented:
> I'm curious if we measured the impact of this change? I didn't think these goroutines would be considered for scheduling until the channel send, so I'm curious where the cost comes from.

I did few tests. 75k ranges on 5 nice e cluster (kV splits 75k). Even without any workload, you could see impact on go scheduler. You are right, of course, that nothing should happen until send. But of course, even without work happening on those ranges, each one still had checkpoints - one every 200ms

erikgrinaker commented:
> > I'm curious if we measured the impact of this change? I didn't think these goroutines would be considered for scheduling until the channel send, so I'm curious where the cost comes from.
> 
> I did few tests. 75k ranges on 5 nice e cluster (kV splits 75k). Even without any workload, you could see impact on go scheduler. You are right, of course, that nothing should happen until send. But of course, even without work happening on those ranges, each one still had checkpoints - one every 200ms

Sure, but I'm specifically talking about the goroutines we're removing in this PR -- do they actually cost anything, or are they ~free? They're only waiting for the final error result from the registration, right, so they shouldn't be scheduled until  the rangefeed terminates?

miretskiy commented:
> Sure, but I'm specifically talking about the goroutines we're removing in this PR -- do they actually cost anything, or are they ~free? They're only waiting for the final error result from the registration, right, so they shouldn't be scheduled until the rangefeed terminates?

I can post some info, while this is getting reviewed. Roughly free is not the same as free, esp since you have so many.  I tested with another PR which removes 1 more go routine on client side, but that change is a lot more disruptive, so I pulled it out for this PR; but I will get updated numbers.
Conceptually, I think this PR should help.
I'm not a go scheduler expert; I do suspect that even though these go routines are not active, all of them are in runnable state; many go routines in runnable state will slow down the scheduler since it has to work harder to select next go routine to run. Fewer go routines is better, in my opinion, even if each one is roughly free

erikgrinaker commented:
> I'm not a go scheduler expert; I do suspect that even though these go routines are not active, all of them are in runnable state; many go routines in runnable state will slow down the scheduler since it has to work harder to select next go routine to run.

I don't think they are. They should be in the blocked state until someone sends on the channel, at which point the sending goroutine will mark them as runnable: https://codeburst.io/diving-deep-into-the-golang-channels-549fd4ed21a8.

miretskiy commented:
> > I'm not a go scheduler expert; I do suspect that even though these go routines are not active, all of them are in runnable state; many go routines in runnable state will slow down the scheduler since it has to work harder to select next go routine to run.
> 
> I don't think they are. They should be in the blocked state until someone sends on the channel, at which point the sending goroutine will mark them as runnable: https://codeburst.io/diving-deep-into-the-golang-channels-549fd4ed21a8.

As I said, not a go scheduler expert. It still seems to me that fewer resources used is better, even if resource is cheap.

miretskiy commented:
I'll post some benches later

Commit:
kvcoord: Add metric tracking number of rangefeed registrations

Add a gauge metric keep track of currently active registrations.

Epic: None

Release note: None

tbg commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,206 @@
+// Make creates Future which will produce the value returned by the produce function.
+// Produce function runs on separate Go routine.  The passed in context should be
+// respected by produce function.
+func Make[T any](ctx context.Context, produce func(ctx context.Context) (T, error)) Future[T] {
```
`Make` is unused and I'm not fond of the fact that it "internalizes" the goroutine (and that it introduces the `Join` method). Such a facility is IMO better kept on the stopper. Can we remove this from the PR to add only what we really need? The `Future` seems useful.

miretskiy replied:
This is gone.

tbg reviewed (commented):
Cursory first review only to get my bearing, thank you for giving this code some attention (I know REPL nominally owns it so double thanks!). Generally speaking I think the introduction of the `Future` makes sense - even if this extra goroutine doesn't weigh heavily on the scheduler, it still seems like a useful pattern to introduce. I'm not sold, however, on the "Promise" which is also not used - I'd much prefer we removed that.

I was wondering if this "flattening" could complicate the work we expect we have to do regardless, to avoid the periodic "pulsing" of the goroutines via the closed timestamp interval. I don't think so, which is good.

Could you list the goroutines before and after in the commit message? I think that would be instructive for most readers including myself. I can see that we're saving a goroutine in `rangeFeedWithRangeID` (previously on `<-errC`). We have `(*registration).outputLoop`. We have the incoming client goroutine, which will now sit on the promise directly (skipping the `<-errC`). And then we have a goroutine in gRPC servicing the streaming RPC.

Is anything conceptually in the way of eliding `outputLoop` as well, by hooking it up directly to the registration? Instead of returning a `Future` to the client goroutine, we'd return - essentially - the `*registration` itself and task the client goroutine with servicing it.

tbg commented:
Ah, I meant to say, curious to see the overhead as well. I played around with this

and got (on my M1 Mac)

```
go run ./pkg/cmd/goroutinepulser 100000 200ms
[...]
avg	p50	p75	p90	p99	p99.9	p99.99	pMax
6.87	6.52	9.65	12.57	p16.69	19.90	21.89	31.46 [cum ms]

```
go run ./pkg/cmd/goroutinepulser 75000 200ms
[...]
avg	p50	p75	p90	p99	p99.9	p99.99	pMax
6.63	6.42	8.98	11.45	p18.17	26.25	29.23	31.46 [cum ms]
```

so here a 25% reduction isn't *dramatic* (plus Erik's point about the goroutines that are being reduced never becoming runnable).

I'm curious to see how a "real" CockroachDB cluster will do.

miretskiy commented:
> Cursory first review only to get my bearing, thank you for giving this code some attention (I know REPL nominally owns it so double thanks!). Generally speaking I think the introduction of the `Future` makes sense - even if this extra goroutine doesn't weigh heavily on the scheduler, it still seems like a useful pattern to introduce. I'm not sold, however, on the "Promise" which is also not used - I'd much prefer we removed that.

I am actually using promise quite extensively in the latest version I just pushed.

> I was wondering if this "flattening" could complicate the work we expect we have to do regardless, to avoid the periodic "pulsing" of the goroutines via the closed timestamp interval. I don't think so, which is good.
> 

I don't think so; but I've been wrong before.

> Could you list the goroutines before and after in the commit message? I think that would be instructive for most readers including myself. I can see that we're saving a goroutine in `rangeFeedWithRangeID` (previously on `<-errC`). We have `(*registration).outputLoop`. We have the incoming client goroutine, which will now sit on the promise directly (skipping the `<-errC`). And then we have a goroutine in gRPC servicing the streaming RPC.
> 

it only removes go routine started by MuxRangefeed to run underlying single rangefeed.

> Is anything conceptually in the way of eliding `outputLoop` as well, by hooking it up directly to the registration? Instead of returning a `Future` to the client goroutine, we'd return - essentially - the `*registration` itself and task the client goroutine with servicing it.

Yes, we should do that.  We start off w/ 5 go routines for regular rangefeed; we are down to 4 w/ Mux; this PR brings it down to 3.
Then, rewrite client portion of mux -- that's another go routine per range.
And then finally, drop output loop.
Everything should move into O(num nodes) instead of O(num ranges).

miretskiy commented:
@erikgrinaker @tbg 
I got some benchmarks.  As you correctly point out, @erikgrinaker , these go routines this PR removes should be roughly
free.  And they are.  But they are not entirely free.  The setup is 5 node, n32-standard machine cluster; KV workload initialized with 100k splits -- so 20k ranges per node.  It's a bit too much, but good for this test to see the impact
of those Go routines.

The biggest savings of this PR come from the startup costs:

<img width="993" alt="Screenshot 2023-02-16 at 2 30 47 PM" src="https://user-images.githubusercontent.com/54641364/219491922-96296055-1552-4266-90e1-c9acb1344373.png">

You can clearly tell a difference in runnable count between up to 160(!) on the left -- this was running master version of MuxRF, and on the right -- which is running this change.

You can also (obviously) see the total number of Go routines go down from 400k to 300k:
<img width="833" alt="Screenshot 2023-02-16 at 4 36 14 PM" src="https://user-images.githubusercontent.com/54641364/219492415-bc5d2d15-66d9-454b-97fe-1c2de27279f9.png">

More interestingly, is the 99.99 impact on sql latency:
<img width="810" alt="Screenshot 2023-02-16 at 4 37 54 PM" src="https://user-images.githubusercontent.com/54641364/219492759-6176ed5c-eeeb-4784-9e1e-459c01ff8d7e.png">

So, to summarize this PR: it eliminates 1 Go routine from MuxRangefeed.  This Go routine used to be idle, so it was mostly free.  However, at the start of rangefeed, that extra go routine wants to run -- and it's this extra Go routine that would make latency worse -- simply by creating more work for the Go scheduler.  

To be very clear: there are other ways to solve it -- perhaps by putting some sort of a rate limit on the creation of those extra go routines.  But again, I just don't see why we needed them in the first place.

tbg commented:
Thanks for running the experiments! Feel free to request a review once the conflicts have been rebased away and it's ready for an in-depth review.

erikgrinaker commented:
Good point, spawning all of these in a loop will put a fair bit of load on the scheduler. I was mostly thinking about the individual rangefeed approach where the goroutine is already spawned by gRPC, not `MuxRangeFeed` where we don't need to spawn them in the first place. Let's get this rebased and wrapped up, and we'll review it. Thanks!

miretskiy commented:
@tbg @erikgrinaker -- rebased/cleaned up.
Ready for review.  I think :)

tbg reviewed (commented):
Looks good! I still need to review the logic in the last commit in a bit more detail but I doubt there will be many surprises.

I did add a bunch of fix-up commits - figured this would be easier than funneling it through a GH review. The largest of them switches to `Future[struct{}]` - that seems like the right one to use here but I may have missed some subtlety.

miretskiy commented on pkg/kv/kvserver/rangefeed/registry.go:
```diff
@@ -13,6 +13,7 @@ package rangefeed
 import (
 	"context"
 	"fmt"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
```
Yes, it is passing lint based on green CI.

miretskiy commented on pkg/kv/kvserver/stores.go:
```diff
@@ -221,10 +222,10 @@ func (ls *Stores) RangeFeed(
 	store, err := ls.GetStore(args.Replica.StoreID)
 	if err != nil {
-		return kvpb.NewError(err)
+		return future.MakeCompletedFuture(kvpb.NewError(err))
```
I'm not sure about `Future[struct{}]` -- but, yes, it's a bit weird with roachpb.Error situation; 
but I think it's more of an exception than a rule.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,205 @@
+
+// Future represents a value which will become ready in the future.
+type Future[T any] struct {
+	// A bit of defensive type embedding in order to have all constructors return
```
I honestly forgot :)
I think at some point, I had a helper `type ValueOr[T] struct {v T, err error}` or something similar, and I wanted to make sure that methods that return future + error, do not returned uninitialized valueOr... but I think at this point, it's moot... So, I need to drop it.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,205 @@
+	mu struct {
+		syncutil.Mutex
+		ready     bool
+		done      atomic.Value // of chan struct{}, created lazily.
```
That's what you get for copying context.Context :)
Removed.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,205 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+//
```
Pretty sure lint was green -- at least based on the CI run; so I think it's fine.

miretskiy commented:
@tbg  -- pushed new version.  I introduced `future.ErrorFuture` -- this seems pretty useful and certainly
more ergonomic vs `future.Future[struct{}]`.  Otherwise, I believe I addressed all of your comments, but apologies if I missed something. Pls take another look.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,205 @@
+	f.runOrDefer(fn, false)
+}
+
+// WhenReadyAsync is similar to WhenReady but executes provided callback on a
```
Method dropped.  Comment updated.

miretskiy commented:
@tbg FYI: I pushed a new version, with largely rewritten `future.Future` implementation -- compliments to a conversation I had with @dt ... The top PR will definitely fail lint, since it's getting a bit too late to fix it for me, but I think new changes in the future library are nice.

In particular, the business with done channel has been reworked.  Instead of accessing this channel under mutex, which 
tends to produce issues when e.g. used in hot loops (same as context.Done), we now have a separate
`type AwaitableFuture[T]` which wraps underlying future to add methods `Get()` to block until value is ready and return it, plus `Done()` -- returning a channel which is always created.  If you only need to use `WhenReady` callbacks, you don't need to create AwaitableFuture, and thus do not incur the cost of mutex lock.
After removal of done channel from base future struct, it became apparent that the only thing remaining that needed
locking was the list of callback functions.  I reworked the implementation to forego the use of mutex altogether in favor of using atomic.Pointer[T] -- the code is a bit more complex, but I think, on the balance, it's much better.

miretskiy commented:
@tbg, @erikgrinaker  -- all is green in the land of CI with all comments addressed.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -1468,36 +1473,40 @@ func (s *lockedMuxStream) Send(e *kvpb.MuxRangeFeedEvent) error {
-	rfGrp := ctxgroup.WithContext(ctx)
-
 	muxStream := &lockedMuxStream{wrapped: stream}
+	var streams syncutil.IntMap
```
Doesn't seem like this is actually used for anything?

miretskiy replied:
You are right; not anymore; though I suspect at some point, some sort of observability around mux spans would be nice. Added a todo for that.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -1468,36 +1473,40 @@ func (s *lockedMuxStream) Send(e *kvpb.MuxRangeFeedEvent) error {
+				})
+				// Sending could fail, but if it did, the stream is broken anyway, so
+				// nothing we can do with this error.
+				_ = sink.Send(&event)
```
Should we close the sink here? That would prevent stuffing subsequent errors or events into it, and also make sure the client isn't left hanging if the send fails for a non-terminal reason.

miretskiy replied:
I don't think sink has a close method.  I could store some additional state in sink, but presumably if Send returned
an error, all subsequent sends ought to return an error as well:

```
// Send blocks until it sends m, the stream is done, or the stream breaks.
// Send must be safe to call on the same stream in different goroutines.
Send(*kvpb.RangeFeedEvent) error
```

erikgrinaker commented on pkg/kv/kvserver/stores.go:
```diff
@@ -206,12 +207,12 @@ func (ls *Stores) SendWithWriteBytes(
+// RangeFeedPromise registers a rangefeed over the specified span. It sends
+// updates to the provided stream and returns a future with an optional error
+// when the rangefeed is complete.
+func (ls *Stores) RangeFeedPromise(
```
nit: I don't think we really need to rename these methods, they still run rangefeeds. I don't particularly care though.

miretskiy replied:
Cool; rename reverted.

erikgrinaker commented on pkg/kv/kvserver/replica_rangefeed.go:
```diff
@@ -341,7 +334,7 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	p := r.rangefeedMu.proc
 	if p != nil {
-		reg, filter := p.Register(span, startTS, catchUpIter, withDiff, stream, errC)
+		reg, filter := p.Register(span, startTS, catchUpIter, withDiff, stream, func() { r.maybeDisconnectEmptyRangefeed(p) }, done)
```
Why aren't we doing `maybeDisconnectEmptyRangefeed()` by attaching a `WhenReady()` to the future? That would avoid having to plumb down yet another callback.

miretskiy replied:
Excellent question.  I almost undid all these changes in favor of WhenReady... Alas, it's not that simple.
There is some annoyance around setting WhenReady on a future that's already ready; this executes fn inline, thus the caller must be careful around mutexes.  Naively, adding done.WhenReady above will deadlock because r.maybeDisconnectEmptyRangefeed will acquire r.rangefeedMu, which is already head.
Annoyances aside, the *real* issue is that this whole registration/unregistration business is very much async.
A signal is sent to the processor to unregister registration.  Crucially, r.maybeDisconnect only disconnects if the number of registration in the processor is 0.  But to get the number of registration, you (async) request processor to return how many registration it has.  All of this is to say: you can't disconnect unless you already un-registred.    However, this future completes before any of the above can take place.  Thus, when maybeDisconnect runs, the count would still be 1.  I've been going back and forth on this, but this is the least invasive way I can come up.

erikgrinaker replied:
Makes sense.

erikgrinaker reviewed (commented):
Basically LGTM, but let's resolve the comments first. Thanks a lot for taking this on!

miretskiy commented on pkg/kv/kvserver/rangefeed/registry.go:
```diff
@@ -403,12 +407,14 @@ func (r registration) String() string {
-func makeRegistry() registry {
+func makeRegistry(n *metric.Gauge) registry {
 	return registry{
+		n:    n,
```
Okay. Passing full metrics struct.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+
+// MakeAwaitableFuture wraps underlying future so that the callers can wait
+// for future to become available.
+func MakeAwaitableFuture[T any](f *Future[T]) AwaitableFuture[T] {
```
Yes; I actually had that done early on.  I happen to really like this pattern.
One of the reasons is that to immediately make awaitable future, you have to construct a channel.  That's not super cheap.  Alternatively, you do something like context does (which Is what I had early) -- and that is, construct channel lazily, but use mutex to synchronize.  That's also not cheap.

I think awaitable future strikes a very good balance.  For one, some users of Future do not want to wait for it -- they just want to arrange for WhenReady to execute when future is ready.

Only the clients that want to wait for future would construct AwaitableFuture, and if the underlying future happens to be completed already (which is fairly common, particularly when using ErrorFuture), the construction is very cheap (since closed channel is used).

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+	return f.done
+}
+
+// WhenReady invokes provided callback when the value (or error) becomes
```
Ack; thanks.  remnants of old api.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+// The caller, of course, is free to specify function that spins up a go
+// routine to work around locking issues.
+func (f *Future[T]) WhenReady(fn func(v T)) {
+	f.runOrDefer(fn)
```
Okay; I used to have different types of callback (I also had Defer() like method that just took func(), but I got rid of it; that's why I had a shared method at some point).  Dito with set (which also had SetErr cousin); but, things got tighter, and shorter, so... All merged now.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+
+// MakePromise constructs future where the caller promises to set the value or
+// an error.
+func MakePromise[T any]() *Future[T] {
```
I don't think these distinctions really matter.  If anything java takes things too far, while
some other implementations do too little...
I would chuck it to some old API remnants (I had a way of constructing "future", which was set by running fn on a go routine.  And thus it was important to distinguish that flavor, from "promise" to set the value... but, that's all gone now)
Going to take you future.Make suggestion to return future, and let java zealots come after me.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+// This method blocks unconditionally.  If the caller needs
+// to respect context cancellation, use Done() method to select
+// on, followed by Get.
+func (f AwaitableFuture[T]) Get() T {
```
It might seem like an anti-pattern, but first, let me say that earlier versions did take ctx, which I eventually removed.

The reason for this is that the use of some future instances, such as a *very important*  `future.Future[error]`, would become ambiguous.  Did this function return context.Cancel because caller context was canceled? Or, did it return context.Cancel because the future was prepared with context.Cancel error?

I think this API is also more consistent with context usage in general.  It has method "Done" and it has method "Err", and it doesn't have a method that combines those.  If the caller wants to combine it, the caller is free to to do so via select.
Furthermore, if you are waiting on a future, it's also quite possible that that you'd want to wait for more than just ctx.Done/future.Done -- you might have other cases you might want to select on.  Having Get method take context would preclude that.

In short, I think your suggestion is *very* reasonable, and that's exactly what I had during earlier iterations; but over time, decided against that.  Does that make sense?

erikgrinaker replied:
>Did this function return context.Cancel because caller context was canceled? Or, did it return context.Cancel because the future was prepared with context.Cancel error?

Isn't this true for all context cancellation errors? They can all mask some other error. Consider e.g. a select clause, which will randomly choose either a context cancellation error or some other error. There are no ordering guarantees.

Besides, the return value would be `(T, error)`, so the caller can distinguish between them if needed.

>I think this API is also more consistent with context usage in general. It has method "Done" and it has method "Err", and it doesn't have a method that combines those.

The crucial distinction is that `Err()` doesn't block, it's a polling API. If `Get()` was a polling API I wouldn't have any objections.

As is, it isn't clear to me that the call in `Node.RangeFeed()` will return on context cancellation. To find that out, I now have to go read/trace all the code in `RangeFeedPromise()` to check that it propagates context cancellation properly via the future and that there are no circular dependencies or deadlocks. If it took a context I wouldn't have to go do that, because it would be obvious, and a dangling future would just get garbage collected.

miretskiy replied:
I split this API into 2.  Get now doesn't wait -- similar to Err; and Wait takes context and returns T, error.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// MakeCompletedErrorFuture constructs completed ErrorFuture.
+func MakeCompletedErrorFuture(err error) *ErrorFuture {
```
Kinda; sorta.  But I still like to keep this.  normally, future.MakeCompletedFuture(err) would work fine.  However, `future.MakeCompletedFuture(nil)` would not.  And neither would anything that's not exactly error, but error-like such as `future.MakeCompletedErrorFuture(kvpb.NewRangeNotFoundError(args.RangeID, s.StoreID()))`

Of course, I could just be more explicit `future.MakeCompletedFuture[error](kvpb.newRangeNotFoundError()...`

If you feel strongly, I can get rid of it.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// Set fulfils the promise and sets the value of the future.
+// Only the first value kept.
```
Been going back and forth on that... Fine, panic it is.  Added a test.

miretskiy commented on pkg/util/future/future.go:
```diff
@@ -0,0 +1,169 @@
+// becomes ready.
+// Use AwaitableFuture if the caller needs to wait for the future.
+type Future[T any] struct {
+	value atomic.Pointer[futureValue[T]]
```
Added; though, you can add when ready callbacks on a future that's already produced.

erikgrinaker replied:
Aren't they called immediately instead of added to the queue?

miretskiy replied:
Sure; but I don't think it needs to be spelled out in the comment on the atomic value.
Do you think what I have so far needs more clarification?

erikgrinaker replied:
Can you push your changes?

miretskiy commented on pkg/util/future/future_test.go:
```diff
@@ -0,0 +1,147 @@
+	}()
+
+	require.Equal(t, 42, future.MakeAwaitableFuture(f).Get())
+	<-ready1
```
done.

miretskiy commented:
@erikgrinaker  -- updates pushed.  Comments addressed, hopefully.

miretskiy commented:
Not really. There is wait method, there is Done. The caller can
differentiate

On Tue, Mar 14, 2023, 7:01 AM Erik Grinaker ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/util/future/future.go
> <https://github.com/cockroachdb/cockroach/pull/96756#discussion_r1135367667>
> :
>
> > +		return AwaitableFuture[T]{Future: f, done: closedChan}
> +	}
> +
> +	done := make(chan struct{})
> +	f.WhenReady(func(v T) {
> +		close(done)
> +	})
> +
> +	return AwaitableFuture[T]{Future: f, done: done}
> +}
> +
> +// Get blocks until future is ready and returns it.
> +// This method blocks unconditionally.  If the caller needs
> +// to respect context cancellation, use Done() method to select
> +// on, followed by Get.
> +func (f AwaitableFuture[T]) Get() T {
>
> Should Get() also return a bool specifying whether it's ready? The zero
> value can be ambiguous, e.g. if storing an int.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/96756#discussion_r1135367667>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVF6ZI6UZR4POHWPTSLW4BFZFANCNFSM6AAAAAAUUSQBTQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

miretskiy commented:
bors r=erikgrinaker

miretskiy commented:
bors r-

miretskiy commented:
bors r+
